kbs/src/plugins/implementations/pkcs11.rs

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) 2025 IBM and Red Hat.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::plugins::resource::{ResourceDesc, StorageBackend};
+use actix_web::http::Method;
+use anyhow::{bail, Context, Result};
+use cryptoki::{
+    context::{CInitializeArgs, Pkcs11},
+    mechanism::Mechanism,
+    object::{Attribute, AttributeInfo, AttributeType, KeyType, ObjectClass},
+    session::{Session, UserType},
+    types::AuthPin,
+};
+use derivative::Derivative;
+use serde::Deserialize;
+use std::{path::PathBuf, sync::Arc};
+use tokio::sync::Mutex;
+use uuid::Uuid;
+
+use super::super::plugin_manager::ClientPlugin;
+
+#[derive(Derivative, Deserialize, Clone, PartialEq)]
+#[derivative(Debug)]
+pub struct Pkcs11Config {
+    /// Path to the PKCS11 module.
+    module: PathBuf,
+
+    /// The index of the slot to be used. If not provided, the first slot will be used.
+    slot_index: Option<u8>,
+
+    /// The user pin for authenticating the session.
+    #[derivative(Debug = "ignore")]
+    pin: String,
+}
+
+pub struct Pkcs11Backend {
+    session: Arc<Mutex<Session>>,
+    wrapkey_id: Uuid,
+}
+
+impl TryFrom<Pkcs11Config> for Pkcs11Backend {
+    type Error = anyhow::Error;
+
+    fn try_from(config: Pkcs11Config) -> anyhow::Result<Self> {
+        let pkcs11 = Pkcs11::new(config.module).context("unable to open pkcs11 module")?;
+        pkcs11.initialize(CInitializeArgs::OsThreads).unwrap();
+
+        let slots = pkcs11.get_slots_with_token()?;
+        let slot_index = usize::from(config.slot_index.unwrap_or(0));
+        if slot_index >= slots.len() {
+            bail!("Slot index out of range");
+        }
+
+        let mut session = pkcs11.open_rw_session(slots[slot_index])?;
+        session.login(UserType::User, Some(&AuthPin::new(config.pin.clone())))?;
+
+        // Generate a UUID to for the wrapping keypair.
+        let wrapkey_id = Uuid::new_v4();
+
+        // Create the HSM wrapping keypair.
+        Pkcs11Backend::wrap_key_new(&mut session, &wrapkey_id)?;
```

Comment:
In the future we may want to do this lazily to avoid using keyslots for people who don't need wrapping keys. Fine for now tho.

Reply:
Agreed. I also want to add a config option if the key already exists on the HSM. The user could just configure the plugin to use a certain keypair.